### PR TITLE
build(edge): extract buildId into environment

### PR DESCRIPTION
**Pull request by huozhi**
### What

* Extract `buildId` and server action encryption key into environment variables for edge to make code more deterministic 
* Fixed the legacy bad env names from #64108
* Always sort `routes` in prerender manifest for consistent output
* Change `environments` to `env` in middleware manifest, confirmed with @javivelasco this is a fine change without need to bumping the version

### Why

Dynamic variants like `buildId`, SA `encryptionKey` and preview props are different per build, which results to the non determinstic edge bundles. Once we extracted them into env vars then the bundles become deterministic which give us more space for optimization

Closes NEXT-3117

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: a9f085d589ac4c2b5824b7ee6476198a165b677c

`__NEXT_EXPERIMENTAL_PPR=true pnpm test-dev test/e2e/switchable-runtime/index.test.ts` (PPR)
- Switchable runtime > Switchable runtime (dev) > should be able to navigate between edge SSR routes without any errors

<details>
<summary>Expand output</summary>

  ● Switchable runtime › Switchable runtime (dev) › should be able to navigate between edge SSR routes without any errors

    TIMED OUT: /to \/edge\/\[id\]/

    <head><meta charset="utf-8"><meta name="viewport" content="width=device-width"><meta name="next-head-count" content="2"><meta name="next-font-preconnect"><noscript data-n-css=""></noscript><link rel="preload" href="/_next/static/chunks/webpack.js" as="script"><link rel="preload" href="/_next/static/chunks/main.js" as="script"><link rel="preload" href="/_next/static/chunks/pages/_app.js" as="script"><link rel="preload" href="/_next/static/chunks/pages/edge/%5Bid%5D.js" as="script"><noscript id="__next_css__DO_NOT_USE__"></noscript></head><body><div id="__next"><div><a href="/edge/foo">to /edge/foo</a></div></div><script src="/_next/static/chunks/react-refresh.js"></script><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{}},"page":"/edge/[id]","query":{},"buildId":"development","nextExport":true,"autoExport":true,"isFallback":false,"customServer":true,"scriptLoader":[]}</script><script nomodule="" src="/_next/static/chunks/polyfills.js"></script><script src="/_next/static/chunks/webpack.js"></script><script src="/_next/static/chunks/main.js"></script><script src="/_next/static/chunks/pages/_app.js"></script><script src="/_next/static/chunks/pages/edge/%5Bid%5D.js"></script><div id="__next-build-watcher" style="position: fixed; bottom: 10px; right: 20px; width: 0px; height: 0px; z-index: 99999;"></div><next-route-announcer><p aria-live="assertive" id="__next-route-announcer__" role="alert" style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; top: 0px; width: 1px; white-space: nowrap; overflow-wrap: normal;"></p></next-route-announcer></body>

    undefined

      642 |
      643 |   if (hardError) {
    > 644 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content + '\n\n' + lastErr)
          |           ^
      645 |   }
      646 |   return false
      647 | }

      at check (lib/next-test-utils.ts:644:11)
      at Object.<anonymous> (e2e/switchable-runtime/index.test.ts:99:9)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js edge/extract-build-id | Change |
| - | - | - | - |
| buildDuration | 14s | 14s | N/A |
| buildDurationCached | 7.6s | 6.4s | N/A |
| nodeModulesSize | 199 MB | 199 MB | ⚠️ +4.79 kB |
| nextStartRea..uration (ms) | 400ms | 408ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge/extract-build-id | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 31.4 kB | 31.4 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | ✓ |
| 8299-HASH.js gzip | 5.1 kB | 5.1 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 242 B | 241 B | N/A |
| main-HASH.js gzip | 29.6 kB | 29.6 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 99 kB | 99 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge/extract-build-id | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge/extract-build-id | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.27 kB | 4.27 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.63 kB | 6.63 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge/extract-build-id | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 485 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge/extract-build-id | Change |
| - | - | - | - |
| index.html gzip | 528 B | 529 B | N/A |
| link.html gzip | 541 B | 542 B | N/A |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge/extract-build-id | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.6 kB | 95.6 kB | N/A |
| page.js gzip | 3.05 kB | 3.05 kB | ✓ |
| Overall change | 3.05 kB | 3.05 kB | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge/extract-build-id | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 624 B | 659 B | N/A |
| middleware-r..fest.js gzip | 155 B | 156 B | N/A |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 839 B | 839 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge/extract-build-id | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | ✓ |
| app-page-exp..prod.js gzip | 97.5 kB | 97.5 kB | ✓ |
| app-page-tur..prod.js gzip | 99.2 kB | 99.2 kB | ✓ |
| app-page-tur..prod.js gzip | 93.5 kB | 93.5 kB | ✓ |
| app-page.run...dev.js gzip | 145 kB | 145 kB | ✓ |
| app-page.run..prod.js gzip | 92 kB | 92 kB | ✓ |
| app-route-ex...dev.js gzip | 21.5 kB | 21.5 kB | ✓ |
| app-route-ex..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 15.2 kB | 15.2 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route.ru...dev.js gzip | 21.1 kB | 21.1 kB | ✓ |
| app-route.ru..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim...dev.js gzip | 23.1 kB | 23.1 kB | ✓ |
| pages.runtim..prod.js gzip | 22.5 kB | 22.5 kB | ✓ |
| server.runti..prod.js gzip | 51.4 kB | 51.4 kB | ✓ |
| Overall change | 948 kB | 948 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | vercel/next.js edge/extract-build-id | Change |
| - | - | - | - |
| 0.pack gzip | 1.58 MB | 1.58 MB | N/A |
| index.pack gzip | 107 kB | 107 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware-b..-manifest.js</strong></summary>

```diff
@@ -3,100 +3,100 @@ self.__BUILD_MANIFEST = {
   devFiles: [],
   ampDevFiles: [],
   lowPriorityFiles: [
-    "static/BUILD_ID/_buildManifest.js",
-    "static/BUILD_ID/_ssgManifest.js",
+    '"static/" + process.env.NEXT_BUILD_ID + "/_buildManifest.js"',
+    '"static/" + process.env.NEXT_BUILD_ID + "/_ssgManifest.js"',
   ],
   rootMainFiles: [
-    "static/chunks/webpack-a3fd1a19c5b09f10.js",
-    "static/chunks/3f784ff6-c1abdf7ce89f5708.js",
-    "static/chunks/2453-8e4d16003c3873d7.js",
-    "static/chunks/main-app-2daec55048c47669.js",
+    "static/chunks/webpack-92f16297e9961465.js",
+    "static/chunks/3c4a14c2-6f82364b9addc180.js",
+    "static/chunks/6901-cc260b1193b8202f.js",
+    "static/chunks/main-app-313341c9dd7f3a76.js",
   ],
   pages: {
     "/": [
-      "static/chunks/webpack-a3fd1a19c5b09f10.js",
+      "static/chunks/webpack-92f16297e9961465.js",
       "static/chunks/framework-4221795723900ad6.js",
-      "static/chunks/main-06e22b173e3e6ebf.js",
-      "static/chunks/pages/index-61e195402d18bbad.js",
+      "static/chunks/main-13668366565bc814.js",
+      "static/chunks/pages/index-c351a6f217e2ff0d.js",
     ],
     "/_app": [
-      "static/chunks/webpack-a3fd1a19c5b09f10.js",
+      "static/chunks/webpack-92f16297e9961465.js",
       "static/chunks/framework-4221795723900ad6.js",
-      "static/chunks/main-06e22b173e3e6ebf.js",
-      "static/chunks/pages/_app-fd0abfb0f9e7c3ac.js",
+      "static/chunks/main-13668366565bc814.js",
+      "static/chunks/pages/_app-6ea050614ac39adf.js",
     ],
     "/_error": [
-      "static/chunks/webpack-a3fd1a19c5b09f10.js",
+      "static/chunks/webpack-92f16297e9961465.js",
       "static/chunks/framework-4221795723900ad6.js",
-      "static/chunks/main-06e22b173e3e6ebf.js",
-      "static/chunks/pages/_error-440da802603da97b.js",
+      "static/chunks/main-13668366565bc814.js",
+      "static/chunks/pages/_error-8d4dd47439ba3a94.js",
     ],
     "/amp": [
-      "static/chunks/webpack-a3fd1a19c5b09f10.js",
+      "static/chunks/webpack-92f16297e9961465.js",
       "static/chunks/framework-4221795723900ad6.js",
-      "static/chunks/main-06e22b173e3e6ebf.js",
-      "static/chunks/pages/amp-d2835241052fddcc.js",
+      "static/chunks/main-13668366565bc814.js",
+      "static/chunks/pages/amp-4c747ca22c01667e.js",
     ],
     "/css": [
-      "static/chunks/webpack-a3fd1a19c5b09f10.js",
+      "static/chunks/webpack-92f16297e9961465.js",
       "static/chunks/framework-4221795723900ad6.js",
-      "static/chunks/main-06e22b173e3e6ebf.js",
+      "static/chunks/main-13668366565bc814.js",
       "static/css/ded6b86ab9cc0a1f.css",
-      "static/chunks/pages/css-c94633659af32fa0.js",
+      "static/chunks/pages/css-0d3dbf0b4197aa76.js",
     ],
     "/dynamic": [
-      "static/chunks/webpack-a3fd1a19c5b09f10.js",
+      "static/chunks/webpack-92f16297e9961465.js",
       "static/chunks/framework-4221795723900ad6.js",
-      "static/chunks/main-06e22b173e3e6ebf.js",
-      "static/chunks/pages/dynamic-3ee3908b591a9891.js",
+      "static/chunks/main-13668366565bc814.js",
+      "static/chunks/pages/dynamic-c7977ad6f692f4c7.js",
     ],
     "/edge-ssr": [
-      "static/chunks/webpack-a3fd1a19c5b09f10.js",
+      "static/chunks/webpack-92f16297e9961465.js",
       "static/chunks/framework-4221795723900ad6.js",
-      "static/chunks/main-06e22b173e3e6ebf.js",
-      "static/chunks/pages/edge-ssr-079ec5f7b0e53e9a.js",
+      "static/chunks/main-13668366565bc814.js",
+      "static/chunks/pages/edge-ssr-40e9455c1a20a49b.js",
     ],
     "/head": [
-      "static/chunks/webpack-a3fd1a19c5b09f10.js",
+      "static/chunks/webpack-92f16297e9961465.js",
       "static/chunks/framework-4221795723900ad6.js",
-      "static/chunks/main-06e22b173e3e6ebf.js",
-      "static/chunks/pages/head-9a76c4e28bd43b5a.js",
+      "static/chunks/main-13668366565bc814.js",
+      "static/chunks/pages/head-24a6068f55a9ba25.js",
     ],
     "/hooks": [
-      "static/chunks/webpack-a3fd1a19c5b09f10.js",
+      "static/chunks/webpack-92f16297e9961465.js",
       "static/chunks/framework-4221795723900ad6.js",
-      "static/chunks/main-06e22b173e3e6ebf.js",
-      "static/chunks/pages/hooks-a721ee829af6b6b5.js",
+      "static/chunks/main-13668366565bc814.js",
+      "static/chunks/pages/hooks-6086caaf8bb85fcc.js",
     ],
     "/image": [
-      "static/chunks/webpack-a3fd1a19c5b09f10.js",
+      "static/chunks/webpack-92f16297e9961465.js",
       "static/chunks/framework-4221795723900ad6.js",
-      "static/chunks/main-06e22b173e3e6ebf.js",
-      "static/chunks/pages/image-0e2eb9d879c090ae.js",
+      "static/chunks/main-13668366565bc814.js",
+      "static/chunks/pages/image-f6d2754c87a32e86.js",
     ],
     "/link": [
-      "static/chunks/webpack-a3fd1a19c5b09f10.js",
+      "static/chunks/webpack-92f16297e9961465.js",
       "static/chunks/framework-4221795723900ad6.js",
-      "static/chunks/main-06e22b173e3e6ebf.js",
-      "static/chunks/pages/link-94077c263115df4c.js",
+      "static/chunks/main-13668366565bc814.js",
+      "static/chunks/pages/link-3fe1e735dff3f921.js",
     ],
     "/routerDirect": [
-      "static/chunks/webpack-a3fd1a19c5b09f10.js",
+      "static/chunks/webpack-92f16297e9961465.js",
       "static/chunks/framework-4221795723900ad6.js",
-      "static/chunks/main-06e22b173e3e6ebf.js",
-      "static/chunks/pages/routerDirect-3eedfbe0cb74c2af.js",
+      "static/chunks/main-13668366565bc814.js",
+      "static/chunks/pages/routerDirect-1ecc94b9914b1970.js",
     ],
     "/script": [
-      "static/chunks/webpack-a3fd1a19c5b09f10.js",
+      "static/chunks/webpack-92f16297e9961465.js",
       "static/chunks/framework-4221795723900ad6.js",
-      "static/chunks/main-06e22b173e3e6ebf.js",
-      "static/chunks/pages/script-566bbcbf5ffe41c3.js",
+      "static/chunks/main-13668366565bc814.js",
+      "static/chunks/pages/script-dbb98a9493fcf110.js",
     ],
     "/withRouter": [
-      "static/chunks/webpack-a3fd1a19c5b09f10.js",
+      "static/chunks/webpack-92f16297e9961465.js",
       "static/chunks/framework-4221795723900ad6.js",
-      "static/chunks/main-06e22b173e3e6ebf.js",
-      "static/chunks/pages/withRouter-50e06258dac406d0.js",
+      "static/chunks/main-13668366565bc814.js",
+      "static/chunks/pages/withRouter-e8dc3efdea425957.js",
     ],
   },
   ampFirstPages: [],
```
</details>
<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: a9f085d589ac4c2b5824b7ee6476198a165b677c

File: packages/next/src/build/webpack/plugins/middleware-plugin.ts [hunk 1]
```diff
@@ -42,10 +42,10 @@ export interface EdgeFunctionDefinition {
   name: string
   page: string
   matchers: MiddlewareMatcher[]
+  environments: Record<string, string>
```
Comment (sokra) [inline packages/next/src/build/webpack/plugins/middleware-plugin.ts:45]: That's a pretty weird name.

```suggestion
  env: Record<string, string>
```

File: packages/next/src/build/webpack/plugins/middleware-plugin.ts [hunk 2]
```diff
@@ -227,7 +227,7 @@ function getCreateAssets(params: {
           name,
           filePath,
         })),
-        environments: opts.edgeEnvironments,
+        env: opts.edgeEnvironments,
```
Comment (ijjk) [inline packages/next/src/build/webpack/plugins/middleware-plugin.ts:230]: Same as https://github.com/vercel/vercel/pull/11549/files#r1591393572

Comment (huozhi) [reply packages/next/src/build/webpack/plugins/middleware-plugin.ts:230]: It was not working properly with #64108, it requires to work with some of the changes from the PR. Consider it was only landed on canary and only changed preview props, we want to do the renaming and fix together

**Review by ijjk — APPROVED** [review]

**Pull request merged**